lib/plugins/aws/provider.js

```diff
@@ -1698,6 +1744,170 @@ class AwsProvider {
     });
   }
+
+  getEcrRepositoryName() {
+    const serviceName = this.serverless.service.getServiceName();
+    const stage = this.getStage();
+    return `serverless-${serviceName}-${stage}`;
+  }
 }
 
+Object.defineProperties(
+  AwsProvider.prototype,
+  memoizeeMethods({
+    dockerLoginToEcr: d(
+      // TODO: REFACTORING
+      async function () {
+        const registryId = await this.getAccountId();
+
+        // TODO: Error handling?
+        const result = await this.request('ECR', 'getAuthorizationToken', {
+          registryIds: [registryId],
+        });
+        const { authorizationToken, proxyEndpoint } = result.authorizationData[0];
+        const decodedAuthToken = Buffer.from(authorizationToken, 'base64').toString().split(':')[1];
+        const dockerArgs = ['login', '--username', 'AWS', '--password-stdin', proxyEndpoint];
+        const dockerLoginPromise = spawnExt('docker', dockerArgs, {
+          stdio: ['pipe', 'inherit', 'inherit'],
+        });
+        dockerLoginPromise.child.stdin.write(decodedAuthToken);
+        dockerLoginPromise.child.stdin.end();
+        await dockerLoginPromise;
+        // TODO: Better information about how credentials are stored?
+      },
+      { promise: true }
+    ),
+    getOrCreateEcrRepository: d(
+      async function () {
+        const repositoryName = this.getEcrRepositoryName();
+        let repositoryUri;
+        try {
+          const result = await this.request('ECR', 'describeRepositories', {
+            repositoryNames: [repositoryName],
+          });
+          repositoryUri = result.repositories[0].repositoryUri;
+        } catch (err) {
+          if (err.providerError && err.providerError.code === 'RepositoryNotFoundException') {
+            // TODO: CONSIDER MORE PARAMS FOR REPO CREATION?
+            const result = await this.request('ECR', 'createRepository', {
+              repositoryName,
+            });
+            repositoryUri = result.repository.repositoryUri;
+          } else {
+            throw err;
+          }
+        }
+        return {
+          repositoryUri,
+          repositoryName,
+        };
+      },
+      { promise: true }
+    ),
+    resolveImageUriAndShaFromPath: d(
+      // TODO: TESTS
+      // TODO: Check for error from spawn calls?
+      async function (imageName, imagePath) {
+        let isDockerfileAvailable = false;
+        const pathToDockerfile = path.join(imagePath, 'Dockerfile');
+
+        try {
+          const stat = await fs.promises.stat(pathToDockerfile);
+          isDockerfileAvailable = stat.isFile();
+        } catch {
+          // pass and handle after catch block
+        }
+        if (!isDockerfileAvailable) {
+          throw new this.serverless.classes.Error(
+            `Could not access "Dockerfile" under path: ${pathToDockerfile}`
+          );
+        }
+
+        const { repositoryUri, repositoryName } = await this.getOrCreateEcrRepository();
+
+        await this.dockerLoginToEcr();
```

Comment:
Those two can probably be issued in parallel

Reply:
Are you talking about `getOrCreateEcrRepository` and `dockerLoginToEcr`? I think it's not safe to issue them in parallel as if the repo doesn't exist, we have to create it before trying to use `dockerLoginToEcr`

Reply:
Hmm.. but aren't we logging in to ECR in general, and not to specific repo?

Or do you mean that if there's no single repo at given region AWS would reject login?

Reply:
You're totally right, please disregard my previous comment